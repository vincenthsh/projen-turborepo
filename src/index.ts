export * from "./turbo-config";
export * from "./turbo";
