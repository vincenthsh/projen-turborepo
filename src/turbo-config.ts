export enum TaskOutputLogsEnum {
  /**
   * 	Displays all logs
   */
  FULL = "full",
  /**
   * Only show the hashes of the tasks
   */
  HASH_ONLY = "hash-only",
  /**
   * Only show logs from cache misses
   */
  NEW_ONLY = "new-only",
  /**
   * Only show logs from task failures
   */
  ERRORS_ONLY = "errors-only",
  /**
   * Hides all task logs
   */
  NONE = "none",
}

export interface TurborepoTaskConfig {
  /**
   * The list of tasks that this task depends on.
   *
   * Prefixing a task in `dependsOn` with a `^` tells `turbo` that this pipeline task depends on the
   * package's topological dependencies completing the task with the `^` prefix first (e.g. "a
   * package's `build` tasks should only run once all of its `dependencies` and `devDependencies`
   * have completed their own `build` commands").
   *
   * Items in `dependsOn` without `^` prefix, express the relationships between tasks at the package
   * level (e.g. "a package's `test` and `lint` commands depend on `build` being completed first").
   */
  readonly dependsOn?: string[];

  /**
   * Defaults to `["dist/**", "build/**"]`. The set of glob patterns of a task's cacheable
   * filesystem outputs.
   *
   * Note: `turbo` automatically logs `stderr`/`stdout` to `.turbo/run-<task>.log`. This file is
   * _always_ treated as a cacheable artifact and never needs to be specified.
   *
   * Passing an empty array can be used to tell `turbo` that a task is a side-effect and thus
   * doesn't emit any filesystem artifacts (e.g. like a linter), but you still want to cache its
   * logs (and treat them like an artifact).
   *
   * @see https://turbo.build/repo/docs/reference/configuration#outputs
   */
  readonly outputs?: string[];

  /**
   * all files in the package that are checked into source control
   *
   * A list of file glob patterns relative to the package's `package.json` to consider when
   * determining if a package has changed. turbo.json is always considered an input.
   *
   * Because specifying an inputs key immediately opts out of the default behavior, you may
   * use the special string `$TURBO_DEFAULT$` within the inputs array to restore turbo's
   * default behavior. This allows you to tweak the default behavior for more granularity.
   *
   * @default []
   * @see https://turbo.build/repo/docs/reference/configuration#inputs
   */
  readonly inputs?: string[];

  /**
   * Whether or not to cache the task `outputs`.
   *
   * Setting `cache` to false is useful for daemon or long-running "watch" or development mode tasks
   * that you don't want to cache.
   *
   * @default true
   * @see https://turbo.build/repo/docs/reference/configuration#cache
   */
  readonly cache?: boolean;

  /**
   * Label a task as persistent to prevent other tasks from depending on long-running processes.
   * Persistent tasks are made interactive by default.
   *
   * Because a long-running process won't exit, tasks that would depend on it would never run.
   * Once you've labeled the task as persistent, turbo will throw an error if other tasks depend on it.
   *
   * This option is most useful for development servers or other "watch" tasks.
   *
   * @default false
   * @see https://turbo.build/repo/docs/reference/configuration#persistent
   */
  readonly persistent?: boolean;

  /**
   * Label a task as `interactive` to make it accept inputs from `stdin` in the terminal UI.
   * Must be used with `persistent`.
   *
   * @default false (Defaults to `true` for tasks marked as `persistent`)
   * @see https://turbo.build/repo/docs/reference/configuration#interactive
   */
  readonly interactive?: boolean;

  /**
   * The list of environment variables a task depends on.
   *
   * @see https://turbo.build/repo/docs/reference/configuration#env
   */
  readonly env?: string[];

  /**
   * An allowlist of environment variables that should be made available to this task's runtime, even
   * when in Strict Environment Mode.
   *
   * @see https://turbo.build/repo/docs/reference/configuration#passthroughenv
   */
  readonly passThroughEnv?: string[];

  /**
   * Set output logging verbosity. Can be overridden by the `--output-logs` CLI option.
   *
   * @default "full"
   * @see https://turbo.build/repo/docs/reference/configuration#outputlogs
   */
  readonly outputLogs?: TaskOutputLogsEnum;
}

export interface TurborepoConfig {
  /**
   * A list of globs for implicit global hash dependencies. The contents of these files will be
   * included in the global hashing algorithm. This is useful for busting the cache based on `.env`
   * files (not in Git) or any root level file that impacts package tasks (but are not represented
   * in the traditional dependency graph (e.g. a root `tsconfig.json`, `jest.config.js`, `.eslintrc`
   * , etc.)).
   *
   * @see https://turborepo.org/docs/reference/configuration#globaldependencies
   */
  readonly globalDependencies?: string[];

  /**
   * A list of environment variables that you want to impact the hash of all tasks. Any change to
   * these environment variables will cause all tasks to miss cache.
   *
   * @see https://turbo.build/repo/docs/reference/configuration#globalenv
   */
  readonly globalEnv?: string[];

  /**
   * A list of environment variables that you want to make available to tasks. Using this key opts
   * all tasks into Strict Environment Variable Mode.
   *
   * Additionally, Turborepo has a built-in set of global passthrough variables for common cases,
   * like operating system environment variables. This includes variables like `HOME`, `PATH`,
   * `APPDATA`, `SHELL`, `PWD`, and more. The full list can be found in the source code.
   *
   * @see https://turbo.build/repo/docs/reference/configuration#globalpassthroughenv
   */
  readonly globalPassThroughEnv?: string[];

  /**
   * An object representing the task dependency graph of your project. `turbo` interprets these
   * conventions to properly schedule, execute, and cache the outputs of tasks in your project.
   *
   * Each key in the tasks object is the name of a task that can be executed by `turbo run`. Turborepo
   * will search the packages described in your Workspace's configuration for `scripts` in `package.json`
   * with the name of the task.
   *
   * Using the rest of the configuration described in the task, Turborepo will run the scripts in the
   * described order, caching logs and file outputs in the outputs key when provided.
   *
   * @see https://turbo.build/repo/docs/reference/configuration#tasks
   */
  readonly tasks?: Record<string, TurborepoTaskConfig>;
}
