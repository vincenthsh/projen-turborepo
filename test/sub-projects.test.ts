import { javascript } from "projen";
import {
  synthProjectSnapshot,
  createSubProject,
  createJSIISubProject,
  createProject,
  parseYaml,
} from "./util";
import { INVALID_PACKAGE_MANAGER_ERROR } from "../src";

describe("TurborepoProject", () => {
  it("should add pnpm-workspace.yaml file", () => {
    expect.assertions(1);

    const project = createProject();

    const subProjectDir = "packages/baz";
    createSubProject({
      parent: project,
      outdir: subProjectDir,
    });

    const synth = synthProjectSnapshot(project);

    expect(parseYaml(synth["pnpm-workspace.yaml"]).packages).toStrictEqual([
      "packages/baz",
    ]);
  });

  it("should add TypeScript path mappings when turned on", () => {
    expect.assertions(2);

    const project = createProject({
      pathMapping: true,
    });

    const subProjectBarDir = "packages/bar";
    const subProjectBar = createSubProject({
      parent: project,
      outdir: subProjectBarDir,
      srcdir: "pages",
    });

    const subProjectBazDir = "packages/baz";
    createSubProject({
      parent: project,
      outdir: subProjectBazDir,
      deps: [subProjectBar.package.packageName],
    });

    const synth = synthProjectSnapshot(project);

    expect(synth["packages/baz/tsconfig.json"].compilerOptions.baseUrl).toBe(
      ".",
    );
    expect(
      synth["packages/baz/tsconfig.json"].compilerOptions.paths,
    ).toStrictEqual({
      [subProjectBar.package.packageName]: ["../bar/pages"],
    });
  });

  it("should add VS Code settings", () => {
    expect.assertions(1);

    const project = createProject({
      name: "@foo/root",
      vscodeMultiRootWorkspaces: true,
    });

    const subProjectDir = "packages/baz";
    createSubProject({
      name: "@foo/sub",
      parent: project,
      outdir: subProjectDir,
    });

    const synth = synthProjectSnapshot(project);

    expect(JSON.parse(synth["@foo-root.code-workspace"])).toMatchSnapshot();
  });

  it("should set composite flag on the tsconfig", () => {
    // expect.assertions(3);

    const project = createProject({ projectReferences: true });

    const subProjectBarDir = "packages/bar";
    const subProjectBar = createSubProject({
      parent: project,
      outdir: subProjectBarDir,
    });
    const subProjectFooDir = "packages/foo";
    const subProjectFoo = createJSIISubProject({
      parent: project,
      outdir: subProjectFooDir,
      deps: [subProjectBar.package.packageName],
    });

    const subProjectBazDir = "packages/baz";
    createSubProject({
      parent: project,
      outdir: subProjectBazDir,
      deps: [
        subProjectFoo.package.packageName,
        subProjectBar.package.packageName,
      ],
    });

    const synth = synthProjectSnapshot(project);

    expect(synth["tsconfig.json"].compilerOptions.composite).toBe(true);
    expect(
      synth[`${subProjectBarDir}/tsconfig.json`].compilerOptions.composite,
    ).toBe(true);
    expect(
      synth[`${subProjectFooDir}/tsconfig.dev.json`].compilerOptions.composite,
    ).toBe(true);
  });

  it("should add TypeScript project references when turned on", () => {
    expect.assertions(1);

    const project = createProject({
      projectReferences: true,
    });

    const subProjectBarDir = "packages/bar";
    const subProjectBar = createSubProject({
      parent: project,
      outdir: subProjectBarDir,
    });

    const subProjectBazDir = "packages/baz";
    createSubProject({
      parent: project,
      outdir: subProjectBazDir,
      deps: [subProjectBar.package.packageName],
    });

    const synth = synthProjectSnapshot(project);

    expect(synth["packages/baz/tsconfig.json"].references).toStrictEqual([
      {
        path: "../bar",
      },
    ]);
  });

  it("should add moduleNameMapper for jest", () => {
    expect.assertions(1);

    const project = createProject({
      jestModuleNameMapper: true,
    });

    const subProjectBarDir = "packages/bar";
    const subProjectBar = createSubProject({
      parent: project,
      outdir: subProjectBarDir,
    });

    const subProjectBazDir = "packages/baz";
    createSubProject({
      parent: project,
      outdir: subProjectBazDir,
      deps: [subProjectBar.package.packageName],
    });

    const synth = synthProjectSnapshot(project);

    expect(
      synth["packages/baz/package.json"].jest.moduleNameMapper,
    ).toStrictEqual({
      [`^${subProjectBar.name}$`]: "<rootDir>/../bar/src",
    });
  });

  it("should have no build job in root for when parallelWorkflows are on", () => {
    expect.assertions(1);

    const project = createProject({
      parallelWorkflows: true,
      mutableBuild: false,
    });

    const subProjectBarDir = "packages/bar";
    const subProjectBar = createSubProject({
      name: "bar",
      parent: project,
      outdir: subProjectBarDir,
    });

    const subProjectBazDir = "packages/baz";
    createSubProject({
      name: "baz",
      parent: project,
      outdir: subProjectBazDir,
      deps: [subProjectBar.package.packageName],
    });

    const synth = synthProjectSnapshot(project);

    expect(synth[".github/workflows/build.yml"]).toMatchSnapshot();
  });

  it("should have one build job in root for when parallelWorkflows are off", () => {
    expect.assertions(1);

    const project = createProject({
      parallelWorkflows: false,
      mutableBuild: false,
    });
    const synth = synthProjectSnapshot(project);

    expect(synth[".github/workflows/build.yml"]).toMatchSnapshot();
  });

  it("should not clobber basePath", () => {
    const baseUrl = "src";

    const project = createProject({
      pathMapping: true,
    });

    const subProjectBarDir = "packages/bar";
    const subProjectBar = createSubProject({
      name: "bar",
      parent: project,
      outdir: subProjectBarDir,
      srcdir: "pages",
    });

    const subProjectBazDir = "packages/baz";
    createSubProject({
      name: "baz",
      parent: project,
      outdir: subProjectBazDir,
      deps: [subProjectBar.package.packageName],
      tsconfig: {
        compilerOptions: {
          baseUrl,
        },
      },
    });

    const synth = synthProjectSnapshot(project);

    expect(synth["packages/baz/tsconfig.json"].compilerOptions.baseUrl).toBe(
      baseUrl,
    );
    expect(
      synth["packages/baz/tsconfig.json"].compilerOptions.paths.bar,
    ).toStrictEqual(["../../bar/pages"]);
  });

  it("should throw error if any of the sub-projects package manager differs from root package manager", () => {
    expect.assertions(1);

    const project = createProject();
    const subProjectBarDir = "packages/bar";
    createSubProject({
      parent: project,
      outdir: subProjectBarDir,
      packageManager: javascript.NodePackageManager.YARN_CLASSIC,
    });

    expect(() => synthProjectSnapshot(project)).toThrowError(
      INVALID_PACKAGE_MANAGER_ERROR,
    );
  });
});
