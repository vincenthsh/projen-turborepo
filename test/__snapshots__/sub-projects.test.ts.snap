// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TurborepoProject should add VS Code settings 1`] = `
Object {
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
  "extensions": Object {
    "recommendations": Array [
      "folke.vscode-monorepo-workspace",
      "q.typescript-mono-repo-import-helper",
    ],
  },
  "folders": Array [
    Object {
      "name": "@foo/root",
      "path": "./",
    },
    Object {
      "name": "@foo/sub",
      "path": "packages/baz",
    },
  ],
  "settings": Object {
    "eslint.workingDirectories": Array [
      "./packages/baz",
    ],
    "jest.disabledWorkspaceFolders": Array [
      "@foo/root",
    ],
  },
}
`;

exports[`TurborepoProject should have no build job in root for when parallelWorkflows are on 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

name: build
on:
  pull_request: {}
  workflow_dispatch: {}
jobs:
  build:
    runs-on: ubuntu-latest
    permissions:
      contents: write
      packages: read
    outputs:
      self_mutation_happened: \${{ steps.self_mutation.outputs.self_mutation_happened }}
    env:
      CI: \\"true\\"
    steps:
      - name: Checkout
        uses: actions/checkout@v4
        with:
          ref: \${{ github.event.pull_request.head.ref }}
          repository: \${{ github.event.pull_request.head.repo.full_name }}
      - name: Setup pnpm
        uses: pnpm/action-setup@v3
        with:
          version: \\"9\\"
      - name: Install dependencies
        run: pnpm i --frozen-lockfile
      - name: build
        run: npx projen build
      - name: Find mutations
        id: self_mutation
        run: |-
          git add .
          git diff --staged --patch --exit-code > .repo.patch || echo \\"self_mutation_happened=true\\" >> $GITHUB_OUTPUT
        working-directory: ./
      - name: Upload patch
        if: steps.self_mutation.outputs.self_mutation_happened
        uses: actions/upload-artifact@v4
        with:
          name: .repo.patch
          path: .repo.patch
          overwrite: true
      - name: Fail build on mutation
        if: steps.self_mutation.outputs.self_mutation_happened
        run: |-
          echo \\"::error::Files were changed during build (see build log). If this was triggered from a fork, you will need to update your branch.\\"
          cat .repo.patch
          exit 1
      - name: Backup artifact permissions
        run: cd dist && getfacl -R . > permissions-backup.acl
        continue-on-error: true
      - name: Upload artifact
        uses: actions/upload-artifact@v4
        with:
          name: build-artifact
          path: dist
          overwrite: true
  turbo:
    name: turbo build
    needs: build
    runs-on: ubuntu-latest
    permissions:
      actions: read
      contents: read
    if: \\"! needs.build.outputs.self_mutation_happened\\"
    steps:
      - name: Download build artifacts
        uses: actions/download-artifact@v4
        with:
          name: build-artifact
          path: dist
      - name: Restore build artifact permissions
        run: cd dist && setfacl --restore=permissions-backup.acl
        continue-on-error: true
      - name: Checkout
        uses: actions/checkout@v4
      - name: Setup turbo cache
        uses: felixmosh/turborepo-gh-artifacts@v3
        with:
          repo-token: \${{ secrets.GITHUB_TOKEN }}
          server-token: \${{ github.sha }}
      - name: Setup pnpm
        uses: pnpm/action-setup@v3
        with:
          version: \\"9\\"
      - name: Install dependencies
        run: pnpm i --frozen-lockfile
      - name: Build
        run: pnpm dlx turbo run build --api=\\"http://127.0.0.1:9080\\" --token=\\"\${{ github.sha }}\\" --team=\\"\${{ github.repository_owner }}\\" --scope=\\"\${{ matrix.scope }}\\" --include-dependencies
    strategy:
      fail-fast: true
      matrix:
        scope:
          - bar
          - baz
"
`;

exports[`TurborepoProject should have one build job in root for when parallelWorkflows are off 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

name: build
on:
  pull_request: {}
  workflow_dispatch: {}
jobs:
  build:
    runs-on: ubuntu-latest
    permissions:
      contents: write
      packages: read
    outputs:
      self_mutation_happened: \${{ steps.self_mutation.outputs.self_mutation_happened }}
    env:
      CI: \\"true\\"
    steps:
      - name: Checkout
        uses: actions/checkout@v4
        with:
          ref: \${{ github.event.pull_request.head.ref }}
          repository: \${{ github.event.pull_request.head.repo.full_name }}
      - name: Setup pnpm
        uses: pnpm/action-setup@v3
        with:
          version: \\"9\\"
      - name: Install dependencies
        run: pnpm i --frozen-lockfile
      - name: build
        run: npx projen build
      - name: Find mutations
        id: self_mutation
        run: |-
          git add .
          git diff --staged --patch --exit-code > .repo.patch || echo \\"self_mutation_happened=true\\" >> $GITHUB_OUTPUT
        working-directory: ./
      - name: Upload patch
        if: steps.self_mutation.outputs.self_mutation_happened
        uses: actions/upload-artifact@v4
        with:
          name: .repo.patch
          path: .repo.patch
          overwrite: true
      - name: Fail build on mutation
        if: steps.self_mutation.outputs.self_mutation_happened
        run: |-
          echo \\"::error::Files were changed during build (see build log). If this was triggered from a fork, you will need to update your branch.\\"
          cat .repo.patch
          exit 1
      - name: Backup artifact permissions
        run: cd dist && getfacl -R . > permissions-backup.acl
        continue-on-error: true
      - name: Upload artifact
        uses: actions/upload-artifact@v4
        with:
          name: build-artifact
          path: dist
          overwrite: true
  turbo:
    name: turbo build
    needs: build
    runs-on: ubuntu-latest
    permissions:
      actions: read
      contents: read
    if: \\"! needs.build.outputs.self_mutation_happened\\"
    steps:
      - name: Download build artifacts
        uses: actions/download-artifact@v4
        with:
          name: build-artifact
          path: dist
      - name: Restore build artifact permissions
        run: cd dist && setfacl --restore=permissions-backup.acl
        continue-on-error: true
      - name: Checkout
        uses: actions/checkout@v4
      - name: Setup turbo cache
        uses: felixmosh/turborepo-gh-artifacts@v3
        with:
          repo-token: \${{ secrets.GITHUB_TOKEN }}
          server-token: \${{ github.sha }}
      - name: Setup pnpm
        uses: pnpm/action-setup@v3
        with:
          version: \\"9\\"
      - name: Install dependencies
        run: pnpm i --frozen-lockfile
      - name: Build
        run: pnpm dlx turbo run build --api=\\"http://127.0.0.1:9080\\" --token=\\"\${{ github.sha }}\\" --team=\\"\${{ github.repository_owner }}\\"
"
`;
